### PR TITLE
feat(jwt): API to provision JWT

### PR DESCRIPTION
**Pull request by wsalles**
closes [PLT-66]

[PLT-66]: https://ydata.atlassian.net/browse/PLT-66?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -43,23 +48,32 @@ func main() {
 
 	logger.Info("Starting: Authentication Service")
 
+	// VaultClient
+	logger.Info("Configuring Vault access 📡")
+	authenticator := coreClients.NewLocalAuthenticator("myroot")
```
Comment (portellaa) [inline cmd/main.go:53]: this should be the `K8sAuthenticator`

File: cmd/main.go [hunk 2]
```diff
@@ -43,23 +48,32 @@ func main() {
...
+	vaultClient, err := coreClients.NewVaultClient(logger, vaultClientConfiguration, "authentication-service-role", authenticator)
```
Comment (portellaa) [inline cmd/main.go:54]: 🤔  this role should be part of the configuration of the `VaultClient` in the go-core

File: internal/controllers/rest_controller.go [hunk 3]
```diff
@@ -32,32 +36,44 @@ func NewRESTController(
...
-	s.Router.GET(rc.configuration.AuthServiceURL, gin.WrapF(rc.CheckForAuthentication))
+	s.Router.GET(rc.configuration.AuthServiceURL, rc.CheckForAuthentication)
 	s.Router.GET(rc.configuration.OIDCCallbackURL, gin.WrapF(rc.OIDCProviderCallback))
 	s.Router.GET(rc.configuration.UserInfoURL, gin.WrapF(rc.UserInfo))
 	s.Router.POST(rc.configuration.LogoutURL, gin.WrapF(rc.Logout))
 
-	s.Router.Any("/:forward", gin.WrapF(rc.CheckForAuthentication))
-	s.Router.Any("/:forward/*any", gin.WrapF(rc.CheckForAuthentication))
+	profileTokensURL := "/auth/profiles/:uuid/tokens"
+	s.Router.GET(profileTokensURL, rc.CheckForAuthentication, rc.GetToken)
```
Comment (portellaa) [inline internal/controllers/rest_controller.go:61]: this method should belong to the group with the token uid in the path, so:

`/auth/profiles/:uuid/tokens/:tokenUUID`

File: internal/controllers/rest_controller.go [hunk 4]
```diff
@@ -32,32 +36,44 @@ func NewRESTController(
...
+	s.Router.POST(profileTokensURL, rc.CheckForAuthentication, rc.CreateToken)
+	s.Router.DELETE(profileTokensURL, rc.CheckForAuthentication, rc.DeleteToken)
```
Comment (portellaa) [inline internal/controllers/rest_controller.go:63]: same as above

File: internal/controllers/rest_controller.go [hunk 5]
```diff
@@ -32,32 +36,44 @@ func NewRESTController(
...
+	synthesizerTokensURL := "/auth/synthesizers/:uuid/tokens"
+	s.Router.GET(synthesizerTokensURL, rc.CheckForAuthentication, rc.GetToken)
```
Comment (portellaa) [inline internal/controllers/rest_controller.go:65]: same as above

File: internal/controllers/rest_controller.go [hunk 6]
```diff
@@ -172,11 +183,99 @@ func (rc RESTController) Logout(w http.ResponseWriter, r *http.Request) {
...
+}
+
+// GetToken returns data from the Vault.
+func (rc RESTController) GetToken(c *gin.Context) {
+	r, w := c.Request, c.Writer
+	resource := strings.Split(r.URL.Path, "/")[2]
+	uuid := c.Params.ByName("uuid")
+	path := fmt.Sprintf("%s/data/%s", resource, uuid)
+
+	data, err := rc.provisionTokenService.Get(path)
```
Comment (portellaa) [inline internal/controllers/rest_controller.go:219]: this is logic for the service, you don't need to split and so on.

you would have a `GetProfileToken`, or in the `Get`, have a type, for example (prefer the first)

File: internal/controllers/rest_controller.go [hunk 7]
```diff
@@ -172,11 +183,99 @@ func (rc RESTController) Logout(w http.ResponseWriter, r *http.Request) {
...
+}
+
+// DeleteToken removes a data from the Vault.
+func (rc RESTController) DeleteToken(c *gin.Context) {
+	r, w := c.Request, c.Writer
+	resource := strings.Split(r.URL.Path, "/")[2]
+	uuid := c.Params.ByName("uuid")
+	path := fmt.Sprintf("%s/data/%s", resource, uuid)
+
+	tokenToBeDeleted := c.Query("token_id")
```
Comment (portellaa) [inline internal/controllers/rest_controller.go:286]: we spoke about this in the slack 😅  let's keep with the original methods

File: internal/services/provision_tokens_service.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+// List returns a data list from the Vault.
+func (pt ProvisionTokens) List(path string) (interface{}, error) {
+	return pt.vaultClient.List(path)
+}
+
+// Create stores data into Vault.
+func (pt ProvisionTokens) Create(path string, ptr models.ProvisionTokenRequest) (models.CustomClaims, error) {
+	if ptr.Name == "" || ptr.Expiration <= 0 {
+		return models.CustomClaims{}, errors.New("an error occurred while provisioning the token")
+	}
```
Comment (portellaa) [inline internal/services/provision_tokens_service.go:46]: which error? can't this be validated by the decoder? 🤔

File: internal/services/provision_tokens_service.go [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+	expirationDay := time.Now().Add(time.Duration(ptr.Expiration) * (time.Hour * 24))
+	tokenID := uuid.New().String()
+	data := VaultData{
+		tokenID: map[string]interface{}{
+			"name":       ptr.Name,
+			"expiration": expirationDay.Unix(),
```
Comment (portellaa) [inline internal/services/provision_tokens_service.go:53]: this is just a comment, not sure what would be best, store the unix timestamp or the iso8601 format

File: internal/services/provision_tokens_service.go [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+	}, nil
+}
+
+// Update stores updated data into Vault.
+func (pt ProvisionTokens) Update(path string, uuid string, data interface{}) error {
+	newData := VaultData{
+		uuid: data,
+	}
+	return pt.vaultClient.Patch(path, newData)
+}
```
Comment (portellaa) [inline internal/services/provision_tokens_service.go:72]: why do we need update? 🤔

**Review by portellaa — CHANGES_REQUESTED** [review]

File: cmd/main.go (hunk 2 above)
Comment (renatoaguimaraes) [reply cmd/main.go:54]: Let's think if we have more than one role. If the role name is moved to go-core, how the client will choose the right one?

File: internal/configurations/rest_controller_configuration.go [hunk 11]
```diff
@@ -12,9 +12,11 @@ type RESTControllerConfiguration struct {
 	AuthServiceURL     string        `envconfig:"AUTHSERVICE_URL" default:"/"`
 	LogoutURL          string        `envconfig:"LOGOUT_URL" default:"/logout"`
 	UserInfoURL        string        `envconfig:"USER_INFO_URL" default:"/userinfo"`
+	UserJWTExpires     time.Duration `envconfig:"USER_JWT_EXPIRES_AT" default:"24"`
```
Comment (renatoaguimaraes) [inline internal/configurations/rest_controller_configuration.go:15]: What 24 means? years, months, weeks, days, hours, minutes, seconds...

File: internal/controllers/rest_controller.go (hunk 5 above)
Comment (renatoaguimaraes) [reply internal/controllers/rest_controller.go:65]: same as above

File: internal/controllers/rest_controller.go (hunk 6 above)
Comment (renatoaguimaraes) [reply internal/controllers/rest_controller.go:219]: Let's use path parameters and define clear paths for each resource/method.

```
GET /profiles/profile-id/tokens
POST /profiles/profile-id/tokens
GET /profiles/profile-id/tokens/token-id
DELETE /profiles/profile-id/tokens/token-id

GET /synthesizers/synthesizer-id/tokens
POST /synthesizers/synthesizer-id/tokens
GET /synthesizers/synthesizer-id/tokens/token-id
DELETE /synthesizers/synthesizer-id/tokens/token-id
```

File: internal/controllers/rest_controller.go [hunk 12]
```diff
@@ -172,11 +183,99 @@ func (rc RESTController) Logout(w http.ResponseWriter, r *http.Request) {
...
+}
+
+// ListTokens returns a data list from the Vault.
+func (rc RESTController) ListTokens(c *gin.Context) {
+	r, w := c.Request, c.Writer
+	resource := strings.Split(r.URL.Path, "/")[2]
+	uuid := c.Params.ByName("uuid")
+	path := fmt.Sprintf("%s/data/%s", resource, uuid)
+
+	data, err := rc.provisionTokenService.List(path)
```
[the remark this answers is not in the document]
Comment (renatoaguimaraes) [reply internal/controllers/rest_controller.go:236]: same as above

File: internal/controllers/rest_controller.go [hunk 13]
```diff
@@ -172,11 +183,99 @@ func (rc RESTController) Logout(w http.ResponseWriter, r *http.Request) {
...
+func (rc RESTController) CreateToken(c *gin.Context) {
+	r, w := c.Request, c.Writer
+
+	var newProvisionToken models.ProvisionTokenRequest
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		rc.internalServerError(w, err)
+		return
+	}
+	json.Unmarshal(reqBody, &newProvisionToken)
```
[the remark this answers is not in the document]
Comment (renatoaguimaraes) [reply internal/controllers/rest_controller.go:256]: agree with @portellaa, let's keep consistent

**Review by renatoaguimaraes — CHANGES_REQUESTED** [review]

File: cmd/main.go (hunk 1 above)
Comment (wsalles) [reply cmd/main.go:53]: ops! hahaha

File: internal/controllers/rest_controller.go (hunk 7 above)
Comment (renatoaguimaraes) [reply internal/controllers/rest_controller.go:286]: Let's use path parameters and define clear paths for each resource/method.

```
GET /profiles/profile-id/tokens
POST /profiles/profile-id/tokens
DELETE /profiles/profile-id/tokens/token-id

GET /synthesizers/synthesizer-id/tokens
POST /synthesizers/synthesizer-id/tokens
DELETE /synthesizers/synthesizer-id/tokens/token-id
```

File: cmd/main.go (hunk 2 above)
Comment (portellaa) [reply cmd/main.go:54]: ok, but don't we configure this in the enviroment variable and assign the role? the client can have just one role, right? 🤔

File: internal/controllers/rest_controller.go (hunk 3 above)
Comment (portellaa) [reply internal/controllers/rest_controller.go:61]: nice catch

File: internal/controllers/rest_controller.go (hunk 7 above)
Comment (portellaa) [reply internal/controllers/rest_controller.go:286]: agreeeee

File: internal/controllers/rest_controller.go [hunk 14]
```diff
@@ -169,31 +195,122 @@ func (rc RESTController) UserInfo(w http.ResponseWriter, r *http.Request) {
...
 	for _, auth := range rc.credentials {
 		token, err := auth.Extract(r)
 		if authErrors.IsTokenNotFound(err) {
 			continue
 		}
-		// credentials sent.
 		rc.logger.Debug(token)
-		return token, nil
+		// credentials sent.
+		return token, reflect.TypeOf(auth).Elem().Name(), nil
```
Comment (renatoaguimaraes) [inline internal/controllers/rest_controller.go:310]: Why do we need to use reflection here and what that information means?

**Review by renatoaguimaraes — CHANGES_REQUESTED** [review]

File: internal/controllers/rest_controller.go [hunk 15]
```diff
@@ -70,14 +87,18 @@ func (rc RESTController) CheckForAuthentication(w http.ResponseWriter, r *http.R
 	// check if the token is expired or signature invalid.
 	// if so, the OIDC flow will be started to recreate a token.
 	if authErrors.IsTokenExpired(err) || authErrors.IsTokenSignatureInvalid(err) {
+		if authType == "HeaderCredentialsHandler" {
+			rc.forbidden(w, err)
+			c.Abort()
+			return
+		}
```
Comment (renatoaguimaraes) [inline internal/controllers/rest_controller.go:94]: I didn't realize this block of code. Can you elaborate a bit the reason behind it?

File: internal/controllers/rest_controller.go [hunk 16]
```diff
@@ -148,16 +166,24 @@ func (rc RESTController) OIDCProviderCallback(w http.ResponseWriter, r *http.Req
...
 	}
 	userInfo, err := rc.oidcService.Decode(token)
+	if authErrors.IsTokenExpired(err) || authErrors.IsTokenSignatureInvalid(err) {
+		if authType == "HeaderCredentialsHandler" {
+			rc.forbidden(w, err)
+			return
+		}
+		rc.logger.Warn(err)
+		rc.RedirectToOIDCProvider(w, r)
+		return
```
Comment (renatoaguimaraes) [inline internal/controllers/rest_controller.go:183]: The same as above, why do we need this authType check?  I think the redirect needs to happen just if the token is expired, an invalid signature just need to return 403.

Comment (renatoaguimaraes) [reply internal/controllers/rest_controller.go:183]: I'm thinking in another use case, for example, machine-to-machine. So , how the redirect will works in that case, when the user is not a web browser.

File: internal/controllers/rest_controller.go [hunk 17]
```diff
@@ -169,31 +195,122 @@ func (rc RESTController) UserInfo(w http.ResponseWriter, r *http.Request) {
...
 	// if a token is not found, return Forbidden.
 	if authErrors.IsTokenNotFound(err) {
-		rc.forbiddenResponse(w, err)
+		rc.forbidden(w, err)
+		return
+	}
+
+	if authType == "HeaderCredentialsHandler" {
+		rc.badRequest(w, errors.New("unable to logout without a cookie"))
+	}
```
Comment (renatoaguimaraes) [inline internal/controllers/rest_controller.go:207]: I didn't get this.

**Review by renatoaguimaraes — CHANGES_REQUESTED** [review]

File: internal/services/provision_tokens_service.go [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+	// ... if exists, delete it.
+	delete(data, tokenID)
+	err = pt.vaultClient.Delete(path)
+	if err != nil {
+		return err
+	}
+	pt.logger.Infof("'%s' token has been deleted.", tokenID)
+
+	for uid, data := range data {
+		err = pt.Update(path, uid, data)
```
Comment (renatoaguimaraes) [inline internal/services/provision_tokens_service.go:96]: The compiler is amazing 😄, but I'm a little bit confused about the context of variable `data`. What do you think to change the name of variable `data` in the loop context?

**Review by renatoaguimaraes — CHANGES_REQUESTED** [review]

File: internal/controllers/rest_controller.go [hunk 19]
```diff
@@ -70,14 +87,18 @@ func (rc RESTController) CheckForAuthentication(w http.ResponseWriter, r *http.R
 	// check if the token is expired or signature invalid.
 	// if so, the OIDC flow will be started to recreate a token.
 	if authErrors.IsTokenExpired(err) || authErrors.IsTokenSignatureInvalid(err) {
+		if authType == "HeaderCredentialsHandler" {
+			rc.forbidden(w, err)
+			c.Abort()
+			return
+		}
 		rc.logger.Warn(err)
 		rc.RedirectToOIDCProvider(w, r)
```
Comment (renatoaguimaraes) [inline internal/controllers/rest_controller.go:96]: we can't redirect when the client is a machine

File: internal/configurations/rest_controller_configuration.go (hunk 11 above)
Comment (wsalles) [reply internal/configurations/rest_controller_configuration.go:15]: hours

File: internal/controllers/rest_controller.go [hunk 20]
```diff
@@ -172,11 +183,99 @@ func (rc RESTController) Logout(w http.ResponseWriter, r *http.Request) {
...
 	}
 
-	rc.deleteSessionCookie(w, "access_token")
+	json.NewEncoder(w).Encode(jsonBody)
+	w.WriteHeader(http.StatusCreated)
+}
+
+// DeleteToken removes a data from the Vault.
+func (rc RESTController) DeleteToken(c *gin.Context) {
+	r, w := c.Request, c.Writer
```
[the remark this answers is not in the document]
Comment (wsalles) [reply internal/controllers/rest_controller.go:281]: ok, i'll remove it.

File: cmd/main.go (hunk 2 above)
Comment (wsalles) [reply cmd/main.go:54]: https://github.com/ydataai/go-core/pull/64/files

Comment (wsalles) [reply cmd/main.go:54]: https://github.com/ydataai/go-core/pull/64

File: internal/controllers/rest_controller.go [hunk 21]
```diff
@@ -172,11 +183,99 @@ func (rc RESTController) Logout(w http.ResponseWriter, r *http.Request) {
...
+
+	expiration := time.Duration(newProvisionToken.Expiration) * time.Hour
+	jsonBody, err := rc.oidcService.Create(data, expiration)
+	if err != nil {
+		rc.internalServerError(w, err)
 		return
 	}
 
-	rc.deleteSessionCookie(w, "access_token")
+	json.NewEncoder(w).Encode(jsonBody)
```
[the remark this answers is not in the document]
Comment (wsalles) [reply internal/controllers/rest_controller.go:275]: I didn't understand this part... is it not to be done that? I mean the `json.NewEncoder(w).Encode(jsonBody)`.

File: internal/controllers/rest_controller.go [hunk 22]
```diff
@@ -172,11 +183,99 @@ func (rc RESTController) Logout(w http.ResponseWriter, r *http.Request) {
...
+		rc.internalServerError(w, err)
+		return
+	}
+	json.Unmarshal(reqBody, &newProvisionToken)
+
+	resource := strings.Split(r.URL.Path, "/")[2]
+	uuid := c.Params.ByName("uuid")
+	path := fmt.Sprintf("%s/data/%s", resource, uuid)
+
+	data, err := rc.provisionTokenService.Create(path, newProvisionToken)
```
[the remark this answers is not in the document]
Comment (wsalles) [reply internal/controllers/rest_controller.go:262]: I didn't understand this part... what did you mean? about the path?

File: internal/controllers/rest_controller.go (hunk 14 above)
Comment (wsalles) [reply internal/controllers/rest_controller.go:310]: to identify what the request is: cookie or authorization header (bearer).

File: internal/controllers/rest_controller.go (hunk 15 above)
Comment (wsalles) [reply internal/controllers/rest_controller.go:94]: to identify what the request is...
if it's an authorization header, it shouldn't redirect to "login" again.

File: internal/controllers/rest_controller.go (hunk 19 above)
Comment (wsalles) [reply internal/controllers/rest_controller.go:96]: this is another case for us to think about it...

File: internal/controllers/rest_controller.go (hunk 17 above)
Comment (wsalles) [reply internal/controllers/rest_controller.go:207]: 🤔  requests like authorization header don't delete cookie...
well... maybe I was wrong...

File: internal/services/provision_tokens_service.go [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+	if !ok {
+		return fmt.Errorf("%s token not found", tokenID)
+	}
+	// ... if exists, delete it.
+	delete(data, tokenID)
+	err = pt.vaultClient.Delete(path)
+	if err != nil {
+		return err
+	}
+	pt.logger.Infof("'%s' token has been deleted.", tokenID)
```
[the remark this answers is not in the document]
Comment (wsalles) [reply internal/services/provision_tokens_service.go:93]: yes and we recreate after that.

File: internal/services/provision_tokens_service.go (hunk 10 above)
Comment (wsalles) [reply internal/services/provision_tokens_service.go:72]: it makes the PATCH, maybe the name is bad.

File: internal/services/provision_tokens_service.go (hunk 9 above)
Comment (wsalles) [reply internal/services/provision_tokens_service.go:53]: so, I prefer Unix... but I'm open to opnions.

File: internal/services/provision_tokens_service.go (hunk 8 above)
Comment (wsalles) [reply internal/services/provision_tokens_service.go:46]: how? i don't know...

File: internal/configurations/rest_controller_configuration.go (hunk 11 above)
Comment (wsalles) [reply internal/configurations/rest_controller_configuration.go]: became explicit again

**Pull request closed without merging**